include/swift/ABI/Metadata.h

```diff
@@ -1577,6 +1577,126 @@ using TupleTypeMetadata = TargetTupleTypeMetadata<InProcess>;
+/// a Swift protocol or (when Objective-C interoperability is enabled) an
+/// Objective-C protocol.
+///
+/// This type always contains a single target pointer, whose lowest bit is
+/// used to distinguish between a Swift protocol referent and an Objective-C
+/// protocol referent.
+template <typename Runtime>
+class TargetProtocolDescriptorRef {
+  using StoredPointer = typename Runtime::StoredPointer;
+  using ProtocolDescriptorPointer =
+    ConstTargetMetadataPointer<Runtime, TargetProtocolDescriptor>;
+
+  enum : StoredPointer {
+    // The bit used to indicate whether this is an Objective-C protocol.
+    IsObjCBit = 0x1U,
+  };
+
+  /// A direct pointer to a protocol descriptor for either an Objective-C
+  /// protocol (if the low bit is set) or a Swift protocol (if the low bit
+  /// is clear).
+  StoredPointer storage;
+
+public:
+  /// Retrieve the protocol descriptor without checking whether we have an
+  /// Objective-C or Swift protocol.
+  /// FIXME: Temporarily public while we roll out TargetProtocolDescriptorRef.
+  ProtocolDescriptorPointer getProtocolDescriptorUnchecked() const {
+    return reinterpret_cast<ProtocolDescriptorPointer>(storage & ~IsObjCBit);
+  }
+
+public:
+  constexpr TargetProtocolDescriptorRef() : storage() { }
+  constexpr TargetProtocolDescriptorRef(nullptr_t) : storage() { }
+
+  TargetProtocolDescriptorRef(
+                        ProtocolDescriptorPointer protocol,
+                        ProtocolDispatchStrategy dispatchStrategy) {
+#if SWIFT_OBJC_INTEROP
+    assert(!protocol ||
+           protocol->Flags.getDispatchStrategy() == dispatchStrategy);
+    storage = reinterpret_cast<StoredPointer>(protocol)
+      | (dispatchStrategy == ProtocolDispatchStrategy::ObjC ? IsObjCBit : 0);
+#else
+    assert(!isObjC);
+    storage = reinterpret_cast<StoredPointer>(protocol);
+#endif
+  }
+
+  explicit constexpr operator bool() const {
+    return storage != 0;
+  }
+
+  /// The name of the protocol.
+  TargetPointer<Runtime, const char> getName() const {
+    return getProtocolDescriptorUnchecked()->Name;
+  }
+
+  /// Determine what kind of protocol this is, Swift or Objective-C.
+  ProtocolDispatchStrategy getDispatchStrategy() const {
+#if SWIFT_OBJC_INTEROP
+    if (isObjC()) {
+      return ProtocolDispatchStrategy::ObjC;
+    }
+#endif
+
+    return ProtocolDispatchStrategy::Swift;
+  }
+
+  /// Determine whether this protocol has a 'class' constraint.
+  ProtocolClassConstraint getClassConstraint() const {
+#if SWIFT_OBJC_INTEROP
+    if (isObjC()) {
+      return ProtocolClassConstraint::Class;
+    }
+#endif
+
+    return getProtocolDescriptorUnchecked()->Flags.getClassConstraint();
+  }
+
+  /// Determine whether this protocol needs a witness table.
+  bool needsWitnessTable() const {
+#if SWIFT_OBJC_INTEROP
+    if (isObjC()) {
+      return false;
+    }
+#endif
+
+    return getProtocolDescriptorUnchecked()->Flags.needsWitnessTable();
```

Comment:
Now that AnyObject is gone, will this ever be false for Swift protocols? Or just maybe in the future?

Reply:
I don't think it will ever be false for Swift protocols, and we could probably drop the flag entirely once I've finished with this pull request. I don't think we'd go the `AnyObject` route again, with a magic protocol that has no witness table... we'd likely add a new kind of generic requirement or flag bit instead.

Reply:
FYI, `needsWitnessTable` is computed based on the dispatch strategy, so it's always `true` for Swift protocols. I've turned this into a `return true;` in a follow-up PR